Objects/bytesobject.c

```diff
@@ -1867,10 +1867,15 @@ bytes_join(PyBytesObject *self, PyObject *iterable_of_bytes)
 }
 
 PyObject *
-_PyBytes_Join(PyObject *sep, PyObject *x)
+PyBytes_Join(PyObject *sep, PyObject *x)
 {
-    assert(sep != NULL && PyBytes_Check(sep));
+    assert(sep != NULL);
     assert(x != NULL);
+    if (!PyBytes_Check(sep)) {
+        PyErr_Format(PyExc_TypeError,
+                     "expected bytes for sep argument, got %T", sep);
+        return NULL;
+    }
```

Comment:
```suggestion
    if (!PyBytes_Check(sep)) {
        PyErr_Format(PyExc_TypeError,
                     "sep: expected bytes instance, %.80s found",
                     Py_TYPE(sep)->tp_name);
        return NULL;
    }
```

May I suggest a similar error as for `PyUnicode_Join` (at least for the separator argument). I think it's `bytes_join` who checks for the type of *x*, so the error would be maybe distinct, but here, if we can, why not having the same error message? (by the way, is '%T' an alias for PyTYPE(sep)->tp_name?, if so, my suggestion can be simplified).

Reply:
I changed the error message. See PEP 737 for %T and not using %.80s.